Added is_convex in sympy.calculus.util

czgdp1807 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References
https://en.wikipedia.org/wiki/Convex_function
 http://www.ifp.illinois.edu/~angelia/L3_convfunc.pdf
https://en.wikipedia.org/wiki/Logarithmically_convex_function
https://en.wikipedia.org/wiki/Logarithmically_concave_function
https://en.wikipedia.org/wiki/Concave_function


#### Brief description of what is fixed or changed
is_convex has been added to sympy.calculus.util
for checking the convexity of univariate continuous
functions. Tests for the same have also been added.

#### Other comments
I am not aware of any such functionality already present in sympy.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* calculus
  * Added is_convex
<!-- END RELEASE NOTES -->

gdp1807 committed:
Added log_concavity and log_convexity in sympy.core.function

Methods for testing of logarithmic concavity and convexity
of univariate expressions have been added to sympy.core.function.

czgdp1807 commented:
@asmeurer Kindly review this PR.

gdp1807 committed:
Added a new line at the end of file

oscargus commented:
It seems like one(?) of your documentation examples give the wrong result. That is why the test fails.

asmeurer commented on sympy/core/function.py:
```diff
@@ -3096,5 +3096,58 @@ def nfloat(expr, n=15, exponent=False):
+    """
+    from sympy import solve_univariate_inequality as s
+
+    syms = args[:]
```
Why don't you just call the argument `*syms`?

czgdp1807 replied:
Yeah sure, I will edit this. :-)

asmeurer commented on sympy/core/function.py:
```diff
@@ -3096,5 +3096,58 @@ def nfloat(expr, n=15, exponent=False):
+
+    var = syms[0]
+    ratio = func * func.diff(var).diff(var) / (func.diff(var)**2)
+    if s(ratio < 1, var):
```
What is `s`?

czgdp1807 replied:
s is `solve_univariate_equality`.

oscargus replied:
I think it may be better to just write the whole thing out. Will increase the readability. (This may have been partly the point of @asmeurer as well.)

czgdp1807 replied:
Sure I will do the same.

czgdp1807 commented:
@oscargus Thanks for bringing that mistake into notice. I will rectify it.
Though is sympy.core.function the right place for `log_concavity` and `log_convexity`?

gdp1807 committed:
Rectified some mistakes

gdp1807 committed:
Added log_concave, log_convex methods

asmeurer commented:
No it isn't. It should go in sympy/functions somewhere. sympy.core.function is only for the base classes for functions.

asmeurer commented:
Or actually sympy.calculus might be more appropriate, since they aren't actually Function subclasses.

czgdp1807 commented:
@asmeurer I will update the PR accordingly. I am thinking to add it in `sympy.calculus`.

gdp1807 committed:
Shifted log_concave and log_convex to sympy.calculus.util

log_concave and log_convex have been shifted to sympy.calculus.util
and corner case of constant functions has also been handled. More
tests have also been added.

gdp1807 committed:
Added references to sympy.calculus.util

czgdp1807 commented:
@asmeurer @oscargus Please review my latest changes.

gdp1807 committed:
Update test_util.py

gdp1807 committed:
Update util.py

oscargus commented:
I do not know for sure, but I guess you should import `log_convex` as `from sympy.calculus.util import log_convex` in the documentation example, and not from `sympy` directly.

asmeurer commented:
It seems rather odd to functions to check for log-convexity but not regular convexity. I wonder if we should also add `convex` or `is_convex`.

asmeurer commented:
This should probably mention in the docstring that it the function must have a second derivative. (log-)convexity in general does not require the function to even be single differentiable. 

Also, (log-)convexity is defined in general on a domain. Some functions are not (log-)convex on the entire real line but they are on some subset. So these function should probably have a domain parameter, which can default to `reals`.

asmeurer commented on sympy/calculus/util.py:
```diff
@@ -569,6 +570,89 @@ def lcim(numbers):
+    var = syms[0]
+    upper_bound = 1
+    if func.diff(var) == 0:
+        ratio = func * func.diff(var).diff(var)
```
You can use `func.diff(var, var)` or `func.diff(var, 2)` to take the second derivative.

asmeurer replied:
In this case, ratio will be 0, no? 

Also, it may be simpler to just write `log(func).diff(var, 2)`. That will also avoid recomputing the first derivative multiple times.

asmeurer commented on sympy/calculus/util.py:
```diff
@@ -569,6 +570,89 @@ def lcim(numbers):
+        return False
+    return True
+
+def log_concave(func, *syms, **kwargs):
```
The logic of these two functions is basically identical. I would simply leave this function's docstring, but remove the code and have it just call `log_convex(-func)`.

asmeurer commented:
Or we could just have a function to check convexity. The user can easily wrap the function in `log` if they want to check for log convexity, since that is basically what the function does anyway.

czgdp1807 commented:
For univariate functions [this document](Lecture Notes – 1 - Nptel
PDFnptel.ac.in › courses › pdf › Module_2) says, that, the univariate functions have non-decreasing slope.

gdp1807 committed:
Added is_convex in sympy.calculus.util

is_convex has been added to sympy.calculus.util
for checking the convexity of univariate continuous
functions. Tests for the same have also been added.

czgdp1807 commented:
@oscargus  What's wrong with my syntax [here](https://travis-ci.org/sympy/sympy/jobs/481230298)?
It worked perfectly when I tested it on my system.

jksuom commented:
```
Compiling sympy/calculus/util.py ...
  File "sympy/calculus/util.py", line 573
    def is_convex(f, *syms, domain = S.Reals):
                                 ^
SyntaxError: invalid syntax
```

czgdp1807 commented:
@jksuom Can you please help with this error? I am unable to understand, how is my `syntax` invalid?

jksuom commented:
It is a restriction on [Python 2](https://docs.python.org/2.7/reference/compound_stmts.html#function-definitions):
```
parameter_list ::=  (defparameter ",")*
                    (  "*" identifier ["," "**" identifier]
                    | "**" identifier
                    | defparameter [","] )
```
Only optional `**kwargs` can follow a starred parameter. There are two ways to fix this: either change the order of `*syms` and `domain=S.Reals` (no spaces around `=`), or make `domain` a keyword parameter in `**kwds`. Its default value can then be implemented as follows:

    domain = kwds.get('domain', S.Reals)

gdp1807 committed:
Removed syntax error

gdp1807 committed:
Update util.py

czgdp1807 commented:
@jksuom Can you please tell me why [this](https://travis-ci.org/sympy/sympy/jobs/481313091) failed? I have experienced this for the first time.

jksuom commented:
It seems that sage tests are having some issue. They sometimes pass and sometimes fail for no obvious reason.

czgdp1807 commented:
Thanks @jksuom for clarification.
@asmeurer @oscargus @jksuom please review the changes. :-)

czgdp1807 commented:
Any changes that are to be made @asmeurer?

oscarbenjamin commented:
It looks as if the tests have failed because of an unrelated issue. If you merge in (or rebase) master they should pass.

oscarbenjamin commented:
You can rebase off master to rerun the tests. Try this (save your work first and then switch to this branch):
```console
    $ git remote add upstream https://github.com/sympy/sympy.git
    $ git fetch upstream
    $ git rebase upstream/master
    $ git push --force
```
The will incorporate the changes from master and then trigger a rerun of the tests.

oscargus commented:
When you rebase, please add a space between `..` and `[X]` (see https://github.com/sympy/sympy/wiki/Writing-documentation ). Not sure if it actually matters, but for consistency. Very nicely written documentation, btw!

oscargus commented on sympy/calculus/tests/test_util.py:
```diff
@@ -161,6 +161,12 @@ def test_lcim():
 
+def test_is_convex():
+
+    assert is_convex(1/x, x, domain = Interval(0, oo)) == True
```
please remove the spaces around = (keyword arguments should not use them).

oscargus commented on sympy/calculus/util.py:
```diff
@@ -569,6 +570,76 @@ def lcim(numbers):
+        The concerned function.
+    syms : Tuple of symbols
+        The variables with respect to which the convexity is to be determined.
+    domain : Interval
```
Add `, optional` after `Interval` and maybe say something about the default value below.

oscargus commented:
I saw some minor things that I commented on, again for consistency, but overall very nice work.

gdp1807 committed:
Improved docs of is_convex

Docs of is_convex has been improved
and the reviews have been taken into
account.

gdp1807 committed:
Merge branch 'concave_convex' of https://github.com/czgdp1807/sympy into concave_convex

czgdp1807 commented:
@oscargus I have taken into account all the changes that you suggested. Thanks for the review. :-)

czgdp1807 commented:
@oscarbenjamin All the tests are passing now. So, I think the work from my side is completed. Waiting for reviews, if any from @asmeurer @oscargus .

oscargus commented:
LGTM
